x/marketmap/types/path.go

```diff
@@ -0,0 +1,213 @@
+package types
+
+import (
+	fmt "fmt"
+	"strings"
+)
+
+// NewPathsConfig returns a new PathsConfig instance. PathsConfig represents
+// the list of convertable markets (paths) that will be used to convert the
+// prices of a set of tickers to a common ticker.
+//
+// For example, if the oracle receives a price for BTC/USDT and USDT/USD, one
+// possible path to get the price of BTC/USD would would be BTC/USDT -> USDT/USD.
+func NewPathsConfig(ticker Ticker, paths ...Path) (PathsConfig, error) {
+	c := PathsConfig{
+		Ticker: ticker,
+		Paths:  paths,
+	}
+
+	if err := c.ValidateBasic(); err != nil {
+		return PathsConfig{}, err
+	}
+
+	return c, nil
+}
+
+// ValidateBasic performs basic validation on the PathsConfig.
+func (c *PathsConfig) ValidateBasic() error {
+	if err := c.Ticker.ValidateBasic(); err != nil {
+		return err
+	}
+
+	if len(c.Paths) == 0 {
+		return fmt.Errorf("at least one path is required for a ticker to be calculated")
+	}
+
+	// Track the routes to ensure that there are no duplicates.
+	routes := make(map[string]struct{})
+	for _, path := range c.Paths {
+		if err := path.ValidateBasic(); err != nil {
+			return err
+		}
+
+		route := path.ShowRoute()
+		if _, ok := routes[route]; ok {
+			return fmt.Errorf("duplicate path found: %s", route)
+		}
+		routes[route] = struct{}{}
+
+		// Ensure that the path ends up converting to the ticker.
+		if !path.Match(c.Ticker.String()) {
+			return fmt.Errorf("path does not match ticker")
+		}
+	}
+
+	return nil
+}
+
+// UniqueTickers returns all of the unique tickers across all of the paths that
+// are part of the PathsConfig. This is particularly useful for determining the
+// set of markets that are required for a given ticker as well as ensuring
+// that a given set of providers can provide the required markets.
+func (c *PathsConfig) UniqueTickers() map[Ticker]struct{} {
+	seen := make(map[Ticker]struct{})
```

Comment:
```suggestion
	seen := make(map[Ticker]struct{}, len(c.Paths))
```

This just pre-allocates the capacity of the map, so that it won't have to resize as we push elements into it.

Reply:
its not exactly clear how many tickers are going to be there so we can pre-allocate the size.